Implement input action duration

Xrayez opened the pull request:
Closes #21005.

A simple mechanism is added to fetch the time in seconds to tell how long an action has been pressed or released. A new timestamp field is added to calculate the difference between current and  action toggle ticks using `OS.get_ticks_usec`.

It means that you can also tell for how long a particular action **has not been pressed**, some gd-pseudocode combining both cases:

```gdscript
# weapon.gd

const CHARGE_TIME = 1.0 # seconds
const RELOAD_TIME = 3.0 # seconds
var reloaded = true

func _physics_process(_delta):
    var duration = Input.get_action_duration("shoot")
    var power = duration / CHARGE_TIME

    if Input.is_action_pressed("shoot") and reloaded:
        if duration > CHARGE_TIME:
            # Maximum power reached, shoot now!
            shoot(power)
            reloaded = false
    else:
        if duration > RELOAD_TIME:
            # Ready to shoot again.
            reloaded = true

    if Input.is_action_just_released("shoot") and reloaded:
        # Shoot with the current power.
        shoot(power)
```

Note: if an action has not been pressed nor released during the game at all, the method will return 0.0 (can also consider returning `-1.0` to indicate that particular case).

Potential limitations: #26828.

### Test project
[input-action-duration.zip](https://github.com/godotengine/godot/files/4256257/input-action-duration.zip)

groud commented:
I am kind of against adding specific things the action system. Getting how long an action has been pressed is far from being a common use case and can easily be worked around with a simple script registering the action states.

I'd rather keep the action system as simple as possible, and avoid adding more logic there unless there is no way around it.

FeatherAntennae commented:
A lot of games make use of such a system. Be it for charging attacks, input/timing based puzzles, jump height, weapon overheating, cooldowns, etc.

Also, a lot of people keep asking about this on pretty much every major engines' support forums.

I agree the Input system should stay as simple as possible, but I personally think such a small addition to the Input system is useful and small enough to outweight the added complexity.

Calinou commented:
The feature seems small and self-contained enough to make this acceptable to me. The diff is only 15 lines after all :slightly_smiling_face:

Calinou reviewed (commented):
Please document the new method in the class reference :slightly_smiling_face:

Xrayez commented:
Added method documentation with some minimal code example with weapon charging mechanics.

groud commented:
Indeed, it is simple and might be used, but it has nothing to do with input to action mapping. Actions are for mapping input, that's all. If we start adding functions to allow game logic there, we are going to end up with a big and complex API I don't want to see. The two problems I see are :
- There are plenty of way to handle timings in a game (input buffers, timers, etc...) 
- what you propose can be done with two lines of code.

IMHO, everything related to timings should be kept out of the action mapping system.

Xrayez commented:
> IMHO, everything related to timings should be kept out of the action mapping system.

Just to clarify some existing logic, there are two methods which have to do with timings currently:

* `is_action_just_pressed` (pressed tick *equals* current tick)
* `is_action_just_released` (!pressed tick *equals* current tick)

Similarly:

* `get_action_duration` (current tick *minus* pressed/released tick)

I believe the first two would be used more often though. Yet you can also use the difference logic to query the same information. Compare:

Just pressed logic:
```gdscript
if is_action_just_pressed("jump"):
    jump()
```
Equivalent code with duration logic:
```gdscript
if Input.is_action_pressed("jump") and is_equal_approx(get_action_duration("jump"), 0.0):
    jump()
```

So if we remove `is_action_just_pressed/released` methods, we can still achieve the same logic with just a few lines of code on the scripting side with `get_action_duration`.

Furthermore, the duration logic can be used to tell whether the action's pressed state was changed (regardless of whether it was pressed or released):

```gdscript
if get_action_duration("shake") == 0.0
    shake()
```

So if anything, it only unifies all the possible use cases rather than adding something specific things to the input system.

groud commented:
Yeah, but this makes things more complex than they need to be. The just_pressed logic is only computed for the current frame and does not involves time-related computations.

If you implement that, we are going to have people asking for how long an action went going unpressed (to implement cooldowns) too, or even a way to get all event's in a list with their respective timstamps, etc... That opens the door to adding a bunch of features that are, in my opinion, out of the scope of the action mapping system.

We have to draw a limit between what belongs to the action mapping system and what does not. For me, we should stop before adding any kind of time-related features. This will ensure the system is kept simple : it maps inputs to actions, that's all.

If we want to handle more complex scenarios, like input buffer, cooldowns, input event pre-shot (I did not find the right name, but I meant when an action is triggered even if pressed too soon. Like a jump registering even if you did not land on the platform yet), we should either move that to another singleton or to a plugin in the assets store.

Xrayez commented:
> If you implement that, we are going to have people asking for how long an action went going unpressed (to implement cooldowns) too

This PR actually covers both pressed and unpressed durations, but I get what you're talking about.

> we should either move that to another singleton or to a plugin in the assets store.

I doubt such a singleton will be implemented in core as you expressed similar concerns about that too in https://github.com/godotengine/godot-proposals/issues/104#issuecomment-575722171:

> I see this proposal as adding yet another layer of complexity over a system that does already have too much: the raw events, the Input singleton storing the state, and the mapping of states to actions.

Exposing direct input states as in #35240 could actually make it much easier to implement plugin-based input time system in my opinion, possibly making this PR somewhat redundant. There's still a hurdle of storing a "database" of action timestamps via script for each and every action which are already available in the engine, it's just that they're not exposed.

Another way to think about this is adding methods which can retrieve internal timestamps to make this easier perhaps:

```gdscript
Input.get_action_physics_frame()
Input.get_action_idle_frame()
```

They don't involve time-related computations as you say, but alleviate the need to create an entire singleton for the only purpose of duplicating action timestamps with `Engine.get_physics/idle_frame` (available to scripting since 3.2: #35092).

Returning to the original purpose of this PR, I still think that the added method is useful for at least of 50% use cases. But I'm not sure what percentage of use cases is required to make this change eligible for merging, and really I'm not even sure if there's a way to know except for the amount of 👍 next to the proposals.

groud commented:
> I doubt such a singleton will be implemented in core as you expressed similar concerns about that too in godotengine/godot-proposals#104 (comment)

The main problem with that is finding a simple but flexible way to do it. Usually, such system is highly dependent on the kind of game you're building, thus I believe the assets store is likely the best way to have this. Nonetheless, if a plugin gets used a lot it could be moved to core afterwards too, that's also a solution we could use.

> There's still a hurdle of storing a "database" of action timestamps via script for each and every action which are already available in the engine, it's just that they're not exposed.

The only thing you need to do that is being able to retrieve the list of actions, which is definitely the simplest way to allow you storing the whole action state.

 > Returning to the original purpose of this PR, I still think that the added method is useful for at least of 50% use cases. But I'm not sure what percentage of use cases is required to make this change eligible for merging, and really I'm not even sure if there's a way to know except for the amount of +1 next to the proposals.

I did not say it wasn't useful, I said it was, IMO, out of the scope of this API. The amount of a +1 is considered, but it has also to be accepted be core developers. This PR will probably need to be discussed in a PR meeting.

groud commented on main/input_default.cpp:
```diff
@@ -135,6 +135,18 @@ float InputDefault::get_action_strength(const StringName &p_action) const {
+	if (!E)
+		return 0.0f;
+
+	if (Engine::get_singleton()->is_in_physics_frame()) {
```
If this gets accepted anyway, this wont work. The number of frames per second in a game might vary, so this computation is definitely not going to output a correct duration if the FPS fluctuate. The timestamp of the action state change should be stored instead to compute the duration.

Eventually, this function could be replaced by a `get_toggle_timestamp()` instead, to retrieve this timestamp directly instead of the duration.

Xrayez replied:
Currently the `Engine` singleton provides the following methods related to FPS:
```cpp
Engine::get_iterations_per_second()
Engine::get_frames_per_second()
// Engine::get_target_fps()
```
From my understanding, `Engine::get_iterations_per_second()` returns the fixed/physics FPS.
If you look below, the idle frame timestamp is used in conjunction with `Engine::get_frames_per_second()` which can fluctuate as you say.

So if we take the fixed FPS, this should output the right duration, am I right?

If we take the idle FPS, then it should compute the average time instead? If that's the case then it should be documented then, because in most cases we're interested in fixed process for logic. That's why I linked #26828 because the engine already seem to have this kind of an issue in the first place.

> The timestamp of the action state change should be stored instead to compute the duration.

The engine already does it, unless you say about exposing the timestamp to script.

Doing this via script won't help either even with `get_toggle_timestamp()` added because the user will still try to implement similar logic for computing duration as in here.

groud replied:
I don't get why you would need all of this, just compute `current_time - time_the_action_changed_state`, that's a lot simpler than trying to use `Engine::get_iterations_per_second()` or `Engine::get_frames_per_second()`.

> So if we take the fixed FPS, this should output the right duration, am I right?

That's not even sure, it might happen the physics frame gets delayed if there are lags. Maybe dropped too, I am not sure.

> The engine already does it, unless you say about exposing the timestamp to script.

It does not look like so. See the `Action` struct in `input_default.h` :
```
	struct Action {
		uint64_t physics_frame;
		uint64_t idle_frame;
		bool pressed;
		float strength;
	};
```
You would just have to store the last update timestamp there, and compute the difference.

Xrayez replied:
Ah I think I get what you mean. You say that the method should just return the difference in frames rather than seconds? That's actually was my initial idea but I thought that it's not as user-friendly as returning the time in seconds, not sure about that.

> It does not look like so. See the Action struct in input_default.h

That's was my first impression too, until I stumbled upon this:

https://github.com/godotengine/godot/blob/dace1ec9126c752792a0566cfd9485a8ad40c0d1/main/input_default.cpp#L425-L440

That's the only part which seems to populate input action states on event parsing. Notice this line of code:
```cpp
is_action_pressed(E->key()) != p_event->is_action_pressed(E->key())
```
It means that the action's state is only updated on change.

So the `Action` struct should actually be more explicit regarding timestamp names (notice `last_*` prefix):
```cpp
struct Action {
	uint64_t last_physics_frame;
	uint64_t last_idle_frame;
	bool pressed;
	float strength;
};
```
And by the way, the `action_state` map is essentially empty if no actions been pressed during the game, so `Input.get_action_duration()` would not actually report the "unpressed" time but `0.0`. I wonder whether it makes sense to "pre-populate" the `action_state` from the `InputMap` on `Input` initialization to help this.

groud replied:
> Ah I think I get what you mean. You say that the method should just return the difference in frames rather than seconds?

No, not at all. I just meant you need to store the timestamp to compute the duration when the user asks it.

> It means that the action's state is only updated on change.

That's why you need to store the timestamp of the last change.

> So the Action struct should actually be more explicit regarding timestamp names.

Those are not timestamps, they correspond to the number of frames/number of physics ticks ran since the beginning of the game. A timestamp should be added as a new field.

Xrayez replied:
I suppose you're talking about using `OS.get_ticks_usec` for timestamps then. But what if a game lags so much that the tick difference gets too big for a single frame?

Here are the changes to the method:
```cpp
float InputDefault::get_action_duration(const StringName &p_action) const {
	const Map<StringName, Action>::Element *E = action_state.find(p_action);
	if (!E)
		return 0;
	return (OS::get_singleton()->get_ticks_usec() - E->get().timestamp) / 1000000.0f;
}
```

`is_equal_approx(Input.get_action_duration("jump"), 0.0)` never reports `True` with this.

So I suppose the current changes are more robust even with possible issues with fluctuating FPS in my opinion.

groud replied:
``get_ticks_usec() is definitely more robust than what you are using right now. Having a peak in a frame is quite rare, and the difference it might cause is at worst the duration of the frame itself, which doesn't even matter. You're solution cannot be accepted in it's current state, as any drop in FPS is going to make the computation completely wrong.

Using is_equal_approx(Input.get_action_duration("jump"), 0.0)` does not make a lot of sense anyway. That's what `is_action_just_pressed() is for.

Xrayez replied:
@groud the only question I have left is: when I push those changes, are you going to approve the PR personally? Yes or no. 🙂

groud replied:
Well, not really. I am still against adding more such logic there, but other contributors might think it's a good idea. What I asked here is only a requirement on the implementation.

Xrayez commented:
Updated the docs to denote that the method may not return accurate results if it's called during the idle frame due to possible FPS fluctuation so it's not recommended to be used for game logic specifically in those cases, yet this is perfectly fine for editor/visual/UI usages for instance.

`OS.get_ticks_usec` timestamps can be used, but it doesn't seem to be as robust compared to computing the duration based on physics and idle frames which also makes it easier/possible to compare against `0.0`.

FeatherAntennae commented:
Okay, this might be a bad idea, I'm still a begginer in C++, I'm not sure of all the implications on memory and performance, and core systems are intimidating, but if the idea is to keep the `Input` singleton as clean and as simple as possible, couldn't we make `Action` an actual class?

What if instead of calling `Input.get_action_just_pressed("shoot")` or anything like that, we were instead just having the singleton expose Actions as something the user can interact with?

The current `Action` struct could be an `ActionState` struct and an Action could be a Class containing an ActionState and the methods like `pressed()`, `just_pressed()`, etc.

The Action itself could contain the logic for timing and the Input singleton would only pass the ActionState to the appropriate Action or fetch an Action from it, exposing it and it's methods to the user.

This way if someone wants to add timing functionalities, a buffer, keep an history or anything else, they could extend Action and make a BufferAction or TimedAction and the Input singleton would stay clean and simple.

The base Action could have the same functionalities as what the current Input singleton exposes and to avoid breaking compatibility, we could keep the functions in the singleton, deprecate them and just replace their content by a call to the same method in the appropriate Action for the time being.

If this wouldn't cause too much problems, I could open a proper proposal for this and try to implement it myself.

Xrayez commented:
> What if instead of calling Input.get_action_just_pressed("shoot") or anything like that, we were instead just having the singleton expose Actions as something the user can interact with?

This somehow reminds me the changes introduced by #36368 where signals can operate as `Callable` objects if I understand correctly (signals are not only represented as strings). Similarly, actions can also be represented as resources which can be passed around.

Some `Action`'s properties like `just_pressed` are time-dependent so if we want the `Action` resource to  not depend on `Engine` singleton calls like `get_physics_frame()`, those need to be updated on each `MainLoop` iteration perhaps. I mean if I were to make `Action extends Resource`, I'd make sure it's purely data-driven. That's where @groud's reasoning makes sense.

With the changes I proposed in godotengine/godot-proposals#104 I went full-blown refactor in #35240 which exposes an entire `InputState` as a resource, which has an additional benefit of keeping not only the timing information connected to actions, but also the global timestamp for the entire input state (keys pressed, joy buttons pressed, mouse motion etc). Things like input buffers would be a matter of duplicating an entire `InputState` on demand and storing an array of those.

The question whether `Input.get_action("jump").pressed` would be more intuitive than `Input.is_action_pressed("jump")`, but having both API is also an option as you said. Some methods like `Input.action_press()` would make more sense as `Input.get_action("jump").pressed = true`.

The `Input` singleton could have methods which operated on a group of actions instead. For instance `Input.get_actions_pressed()` #15681, which would return a list of `Action` resources, each having necessary data to tell specific things about them. Sort of like `get_tree.get_nodes_in_group()` but for actions. In fact you could go further and assign specific controller ids to each type of actions perhaps.

There are many fruitful possibilities. Some would say that it adds complexity, but to me that's flexibility (which requires maintenance efforts by other core contributors unfortunately because there's only a handful of them).

Xrayez commented:
Alright I made changed as suggested by @groud. Instead of relying on physics and idle frames, `OS.get_ticks_usec` timestamp is added to Action struct.

Some pros and cons I personally find with this:

Pros:
* Doesn't depend on FPS which can fluctuate on idle time (physics time should be fixed but as @groud said it may drop a few frames occasionally), so avoids inaccurate results with great fluctuations.
* Higher granularity/precision for when an action was *physically* pressed, not only frame-by-frame duration.

Cons:
* Since the timestamp may not represent the logical game time, it might be more difficult to do equality checks (even with `is_equal_approx(get_action_duration("jump"), 0.0)`).
* Recording and later replaying events might not be as accurate due to physical time vs physics frame time divergences, but this isn't critical for most use cases anyway I guess.

groud commented on main/input_default.cpp:
```diff
@@ -430,6 +437,7 @@ void InputDefault::_parse_input_event_impl(const Ref<InputEvent> &p_event, bool
 				Action action;
 				action.physics_frame = Engine::get_singleton()->get_physics_frames();
 				action.idle_frame = Engine::get_singleton()->get_idle_frames();
+				action.timestamp = OS::get_singleton()->get_ticks_usec();
```
This still cannot work, you have to update the timestamp only when the pressed value changed. Otherwise, when only the strength changes (for axes), the duration will be modified too.

Xrayez replied:
Just tested, the timestamp is not modified whenever axes (strength) changes (only when it goes to 0 or 1). So works on my machine. :eyes:

See updated test project in the PR description.

groud replied:
Ah sorry, I did not see the check for changing pressed status at the top. You are right :)

Commit:
Implement input action duration

A simple mechanism is added to fetch the time in seconds to tell how
long an action has been pressed or released. A new timestamp field is
added to calculate the difference between current and action toggle
ticks using OS::get_ticks_usec.

Xrayez commented:
Rebased following the `InputFilter` renamings.

I'm not sure about the direction of this PR. I wanted to close this at some point given my findings about the [hidden ideology regarding the development of Godot](https://github.com/godotengine/godot-proposals/issues/575#issuecomment-623128660) and how allegedly specific features like this does not belong in core, but I'll leave this PR up to community anyways.

To further decrease the chances of this being merged (😄 ) I wanted to say the following, please don't feel offended because that's just my experience:

@groud from the first moment I started contributing to Godot (like 3 years ago), one of the first things I received is your comments on how my features are too specific and outright shadowing the intrinsic value of a feature being proposed. While I understand that you want to ensure that the engine remains healthy in terms of preventing the software bloat, this attitude is *not* healthy towards people who do try to improve the engine. I realize this is just me with my emotional issues but I had to say that.

Cheers. ❤️

Xrayez commented:
So far it seems like most my proposals are on the hard line of engine vs user responsibility, so that's where I'm torn oftentimes.

The discussion takes me quite a bunch of personal energy (not to mention the increased anxiety this process involves for me, especially at #godotengine-devel IRC). Besides, the work I do on the engine can usually be streamlined downstream without much issues, as I already have a fork which is specific to my project's requirements.

> there is nothing we can really do about it as the project becomes bigger over time.

godotengine/godot-proposals#575.

80-90% of pull requests like this one can be prevented by having a more or less clear criteria of what goes into the engine vs plugin. Only the remaining 10-20% are worth discussing at #godotengine-devel.

Yeah there's a link at CONTRIBUTING.md document which points [to this article](https://godotengine.org/article/will-your-contribution-be-merged-heres-how-tell) though, but I guess the visibility of the decision diagram there asks for the better IMO (it's only an article, doesn't have much enforcement on the documentation side, so developers may ignore it).

Also, I'm totally fine if even 50% of pull requests will be closed, as long as I can advertise my skills to general public that way, hehe.

All I'm asking is a balance between the critique for the sake of it (so it doesn't become a habit), and having ability to see the other side of the coin. 😛

Xrayez commented:
Closing this for now unless core devs express explicit interest in this and I'll reopen accordingly, I don't see the point of having this in Limbo, cheers. 😃

There's another PR which perhaps is more in alignment with the scope of the `Input` class: #35012.

Xrayez commented:
Reopening because this can help implementing goostengine/goost#16, likely not in `3.2` but certainly in the future.

To resurrect the discussion and regurgitate my points a bit regarding the scope of the proposed changes: we already have time-dependent `is_action_just_pressed` and `is_action_just_released`, so makes sense to add something similar which actually covers both those methods: `get_action_duration`, see above discussion!
